### ISSUE TITLE
[BUG] Error: backup-target is empty with NFS (when it's apparently not)

### ISSUE THREAD

**Issue by immanuelfodor** [longhorn/longhorn#2392]
**Describe the bug**
The UI says "error listing backups: setting backup-target is empty" when an NFS target is specified on a fresh Longhorn installation. The setting is present in the `ConfigMap` and even the UI shows it on the Settings page, so it shouldn't be throwing the error.

![image](https://user-images.githubusercontent.com/21174107/111905988-c27d3600-8a4e-11eb-8e2a-ce2ce33ff29c.png)

![image](https://user-images.githubusercontent.com/21174107/111906388-a7abc100-8a50-11eb-9dc5-a70374cd4283.png)

**To Reproduce**
1. Install Longhorn:
```bash
helm repo add longhorn https://charts.longhorn.io
helm repo update
kubectl create ns longhorn-system
helm install longhorn longhorn/longhorn \
      --namespace longhorn-system \
      --set defaultSettings.defaultDataLocality=best-effort \
      --set defaultSettings.nodeDownPodDeletionPolicy=delete-both-statefulset-and-deployment-pod \
      --set persistence.defaultClass=false \
      --set defaultSettings.backupTarget=nfs://192.168.1.1:/nfs/longhorn-backup
kubectl apply -f servicemonitor.yml  # from https://longhorn.io/docs/1.1.0/monitoring/integrating-with-rancher-monitoring/#add-longhorn-metrics-to-the-rancher-monitoring-system
```
2. Open the Longhorn UI (both through Rancher web proxy or `kubectl port-forward` just to make sure it's not a proxy issue)
3. Click the Backup page
4. See the error, all buttons are disabled
5. See the error in the manager pod logs
6. Click the Settings page, see the setting present
7. Check the ConfigMap, see the setting present

**Expected behavior**
The backup page doesn't show an error and activates disabled functionality according to the setting YAML.

**Log**

```
$ k logs longhorn-manager-747ms | tail -n 16
192.168.1.2 - - [21/Mar/2021:13:06:43 +0000] "GET /v1/backupvolumes HTTP/1.1" 500 209 "" "Mozilla/5.0 (..."
time="2021-03-21T13:06:43Z" level=warning msg="HTTP handling error error listing backups: setting backup-target is empty"
time="2021-03-21T13:06:43Z" level=error msg="Error in request: error listing backups: setting backup-target is empty"
time="2021-03-21T13:06:43Z" level=warning msg="HTTP handling error error listing backups: setting backup-target is empty"
time="2021-03-21T13:06:43Z" level=error msg="Error in request: error listing backups: setting backup-target is empty"
192.168.1.2 - - [21/Mar/2021:13:06:43 +0000] "GET /v1/backupvolumes HTTP/1.1" 500 209 "" "Mozilla/5.0 (..."
time="2021-03-21T13:06:43Z" level=warning msg="HTTP handling error error listing backups: setting backup-target is empty"
time="2021-03-21T13:06:43Z" level=error msg="Error in request: error listing backups: setting backup-target is empty"
192.168.1.2 - - [21/Mar/2021:13:06:43 +0000] "GET /v1/backupvolumes HTTP/1.1" 500 209 "" "Mozilla/5.0 (..."
10.42.2.136 - - [21/Mar/2021:13:07:01 +0000] "GET /metrics HTTP/1.1" 200 838 "" "Prometheus/2.18.2"
10.42.2.136 - - [21/Mar/2021:13:08:01 +0000] "GET /metrics HTTP/1.1" 200 840 "" "Prometheus/2.18.2"
10.42.2.136 - - [21/Mar/2021:13:09:01 +0000] "GET /metrics HTTP/1.1" 200 839 "" "Prometheus/2.18.2"
10.42.2.136 - - [21/Mar/2021:13:10:01 +0000] "GET /metrics HTTP/1.1" 200 838 "" "Prometheus/2.18.2"
10.42.2.136 - - [21/Mar/2021:13:11:01 +0000] "GET /metrics HTTP/1.1" 200 842 "" "Prometheus/2.18.2"

$ helm template longhorn longhorn/longhorn --namespace longhorn-system --set defaultSettings.defaultDataLocality=best-effort --set defaultSettings.nodeDownPodDeletionPolicy=delete-both-statefulset-and-deployment-pod --set persistence.defaultClass=false --set defaultSettings.backupTarget=nfs://192.168.1.1:/nfs/longhorn-backup | yq r - -d 2
# Source: longhorn/templates/default-setting.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: longhorn-default-setting
  namespace: longhorn-system
  labels:
    app.kubernetes.io/name: longhorn
    helm.sh/chart: longhorn-1.1.0
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/instance: longhorn
    app.kubernetes.io/version: v1.1.0
data:
  default-setting.yaml: "backup-target: nfs://192.168.1.1:/nfs/longhorn-backup\nbackup-target-credential-secret: \nallow-recurring-job-while-volume-detached: \ncreate-default-disk-labeled-nodes: \ndefault-data-path: \nreplica-soft-anti-affinity: \nstorage-over-provisioning-percentage: \nstorage-minimal-available-percentage: \nupgrade-checker: \ndefault-replica-count: \ndefault-data-locality: best-effort\nguaranteed-engine-cpu: \ndefault-longhorn-static-storage-class: \nbackupstore-poll-interval: \ntaint-toleration: \npriority-class: \nauto-salvage: \nauto-delete-pod-when-volume-detached-unexpectedly: \ndisable-scheduling-on-cordoned-node: \nreplica-zone-soft-anti-affinity: \nvolume-attachment-recovery-policy: \nnode-down-pod-deletion-policy: delete-both-statefulset-and-deployment-pod\nallow-node-drain-with-last-healthy-replica: \nmkfs-ext4-parameters: \ndisable-replica-rebuild: \nreplica-replenishment-wait-interval: \ndisable-revision-counter: \nsystem-managed-pods-image-pull-policy: \nallow-volume-creation-with-degraded-availability: \nauto-cleanup-system-generated-snapshot:"

$ k exec longhorn-manager-747ms -- cat /var/lib/longhorn-setting/default-setting.yaml
backup-target: nfs://192.168.1.1:/nfs/longhorn-backup
backup-target-credential-secret: 
allow-recurring-job-while-volume-detached: 
create-default-disk-labeled-nodes: 
default-data-path: 
replica-soft-anti-affinity: 
storage-over-provisioning-percentage: 
storage-minimal-available-percentage: 
upgrade-checker: 
default-replica-count: 
default-data-locality: best-effort
guaranteed-engine-cpu: 
default-longhorn-static-storage-class: 
backupstore-poll-interval: 
taint-toleration: 
priority-class: 
auto-salvage: 
auto-delete-pod-when-volume-detached-unexpectedly: 
disable-scheduling-on-cordoned-node: 
replica-zone-soft-anti-affinity: 
volume-attachment-recovery-policy: 
node-down-pod-deletion-policy: delete-both-statefulset-and-deployment-pod
allow-node-drain-with-last-healthy-replica: 
mkfs-ext4-parameters: 
disable-replica-rebuild: 
replica-replenishment-wait-interval: 
disable-revision-counter: 
system-managed-pods-image-pull-policy: 
allow-volume-creation-with-degraded-availability: 
auto-cleanup-system-generated-snapshot:

$ k describe pod longhorn-manager-747ms | grep Image
    Image:         longhornio/longhorn-manager:v1.1.0
    Image ID:      docker-pullable://longhornio/longhorn-manager@sha256:3917a6835015d0e5275ef111221af0306c695090204a010865bf992762247889
```

**Environment:**
 - Longhorn version: latest installed by Helm (appVersion: v1.1.0, version: 1.1.0)
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE 
```
$ rke version
INFO[0000] Running RKE version: v1.2.6                  
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-18T16:03:00Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 3 (same as master)
 - Node config
   - OS type and version: CentOS 8 Stream
   - CPU per node: 6
   - Memory per node: 8G
   - Disk type(e.g. SSD/NVMe): NVMe
   - Network bandwidth between the nodes: 10G (Linux bridge with same subnet on the same hypervisor, also true for NFS server)
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): KVM (Proxmox)
 - Number of Longhorn volumes in the cluster: 0 (just installed Longhorn and opened the UI)

**Additional context**
Maybe related to https://github.com/longhorn/longhorn/issues/1382 or https://github.com/longhorn/longhorn/issues/1468 ? I couldn't find any helpful information.

NFS server: NFSv4.2, the share is `/nfs` and `longhorn-backup` is a folder on the share (also tried without creating the folder first and also with 777 permission then deleting all Longhorn pods just to make sure). I don't think the problem is with the NFS server, the NFS subdir external provisioner is working fine over the same server. 

```
$ showmount -e 192.168.1.1
Export list for 192.168.1.1:
/nfs 192.168.1.0/24
```

**Comment by jenting** [comment]
@immanuelfodor
Thank you for the reporting, however, I can't reproduce it. Have you tried to uninstall the helm chart and re-install it again?

**Comment by immanuelfodor** [comment]
I didn't try that but I can, sure. What will happen to an existing volume, will it be preserved or completely destroyed upon the uninstall? Now I have one attached to a pod (with 3 replicas). 

When I try to backup the volume, the snapshot is created fine but the backup fails with the same backup-target missing error. Of course, I can manually save it to a hostPath volume from within a "migration/rescue" pod to aviod data loss.

**Comment by jenting** [comment]
It will be completely destroyed.

**Comment by immanuelfodor** [comment]
After uninstall and install, I get a different error on the Backup page:

`error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only nfs://192.168.1.1:/nfs/longhorn-backup], output Cannot mount nfs 192.168.1.1:/nfs/longhorn-backup: nfsv4 mount failed but nfsv3 mount succeeded, may be due to server only supporting nfsv3: Failed to execute: mount [-t nfs4 -o nfsvers=4.2 192.168.1.1:/nfs/longhorn-backup /var/lib/longhorn-backupstore-mounts/192_168_1_1/nfs/longhorn-backup], output mount.nfs4: mounting 192.168.1_1:/nfs/longhorn-backup failed, reason given by server: No such file or directory , error exit status 32 , stderr, time="2021-03-22T07:05:39Z" level=error msg="Cannot mount nfs 192.168.1.1:/nfs/longhorn-backup: nfsv4 mount failed but nfsv3 mount succeeded, may be due to server only supporting nfsv3: Failed to execute: mount [-t nfs4 -o nfsvers=4.2 192.168.1.1:/nfs/longhorn-backup /var/lib/longhorn-backupstore-mounts/192_168_1_1/srv/rke-nfs/longhorn-backup], output mount.nfs4: mounting 192.168.1.1:/nfs/longhorn-backup failed, reason given by server: No such file or directory\n, error exit status 32" , error exit status 1`

The funny thing is that the backup folder is on the share with 777 and I have other pods using the same share over 4.2 with the NFS provisioner. On one node: 

```
$ mount | grep nfs
192.168.1.1:/harbor-data-harbor-harbor-redis-0-pvc-aab27a04-c828-44bc-8051-3ef2fe03ca75 on /var/lib/kubelet/pods/6a5b4cf9-339f-4382-89c7-d6d29644388e/volumes/kubernetes.io~nfs/pvc-aab27a04-c828-44bc-8051-3ef2fe03ca75 type nfs4 (rw,relatime,vers=4.2,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.1.3,local_lock=none,addr=192.168.1.1)
```

Update: just as I copied this here, I noticed the share paths are different, modified to /longhorn-backup and now it works, no errors on the UI. Since /nfs is the root, it shouldn't be there when using NFSv4. So a helm reinstall and the path adjustment were both needed. Thanks for the help!
